librocksdb_sys/crocksdb/c.cc

```diff
@@ -3432,6 +3432,20 @@ crocksdb_table_properties_t* crocksdb_compactionfiltercontext_table_properties(
       .get();
 }
 
+const char* crocksdb_compactionfiltercontext_start_key(
+    crocksdb_compactionfiltercontext_t* context) {
+  char* result = nullptr;
+  result = CopyString(context->rep.start_key.ToString());
```

Comment:
Seems there is an internal allocation. So where to free it?

Reply:
You can return (ptr, length) and combine them into a `&str` in Rust.

Reply:
@MrCroxx You can check implementation of `crocksdb_sst_partitioner_context_smallest_key`

Reply:
I imitated function `crocksdb_get`:

```cpp
char* crocksdb_get(crocksdb_t* db, const crocksdb_readoptions_t* options,
                   const char* key, size_t keylen, size_t* vallen,
                   char** errptr) {
  char* result = nullptr;
  std::string tmp;
  Status s = db->rep->Get(options->rep, Slice(key, keylen), &tmp);
  if (s.ok()) {
    *vallen = tmp.size();
    result = CopyString(tmp);
  } else {
    *vallen = 0;
    if (!s.IsNotFound()) {
      SaveError(errptr, s);
    }
  }
  return result;
}
```

It seems the function has the same problem. 

> You can return (ptr, length) and combine them into a &str in Rust.

I think this is a better way. I will change to this method.

Reply:
@tabokie @hicqu I remember that `Slice` in rocksdb is not the same as normal c string. rocksdb may use `varint` to encode int when writing to `Slice`.  Will directly using `Slice.data` like `crocksdb_sst_partitioner_context_smallest_key` to construct `&str` lead to a wrong string data? Is it better to use the result of `Slice.ToString()` as `&str`?

Reply:
Here I made a mistake. What you need in Rust is `&[u8]` instead of `&str`.

Reply:
Forget it, I check `Slice::ToString`, it just copy and wrap the raw data.

```cpp
// Return a string that contains the copy of the referenced data.
std::string Slice::ToString(bool hex) const {
  std::string result;  // RVO/NRVO/move
  if (hex) {
    result.reserve(2 * size_);
    for (size_t i = 0; i < size_; ++i) {
      unsigned char c = data_[i];
      result.push_back(toHex(c >> 4));
      result.push_back(toHex(c & 0xf));
    }
    return result;
  } else {
    result.assign(data_, size_);
    return result;
  }
}
```